packages/types/index.ts

```diff
@@ -0,0 +1,161 @@
+
+}
+
+/**
+ * Binds a quad pattern term's position to a variable.
+ */
+export type PatternBindings = Record<string, RDF.Variable>;
+
+export interface IActionQueryOperation extends IAction {
+  /**
+   * The query operation to handle.
+   */
+  operation: Algebra.Operation;
+}
+
+/**
+ * Query operation output.
+ * @see IActorQueryOperationOutputBindings, IActorQueryOperationOutputQuads, IActorQueryOperationOutputBoolean
+ */
+export type ActorQueryOperationOutput =
+  IActorQueryOperationOutputStream |
+  IActorQueryOperationOutputQuads |
+  IActorQueryOperationOutputBoolean;
+export interface IActorQueryOperationOutputBase {
+  /**
+   * The type of output.
+   */
+  type: string;
+  /**
+   * The resulting action context.
+   */
+  context?: ActionContext;
+}
+
+/**
+ * Super interface for query operation outputs that represent some for of stream.
+ * @see IActorQueryOperationOutputBindings, IActorQueryOperationOutputQuads
+ */
+export interface IActorQueryOperationOutputStream extends IActorQueryOperationOutputBase {
+  /**
+   * Callback that returns a promise that resolves to the metadata about the stream.
+   * This can contain things like the estimated number of total stream elements,
+   * or the order in which the bindings appear.
+   * This callback can be invoked multiple times.
+   * The actors that return this metadata will make sure that multiple calls properly cache this promise.
+   * Metadata will not be collected until this callback is invoked.
+   */
+  metadata?: () => Promise<Record<string, any>>;
+}
+
+/**
+ * Data interface for the type of action.
+ */
+export interface IAction {
+
+  /**
+   * The optional input context that is passed through by actors.
+   */
+  context?: ActionContext;
+}
+
+/**
+ * An immutable key-value mapped context that can be passed to any (@link IAction}.
+ * All actors that receive a context must forward this context to any actor, mediator or bus that it calls.
+ * This context may be transformed before forwarding.
+ *
+ * Each bus should describe in its action interface which context entries are possible (non-restrictive)
+ * and expose a `KEY_CONTEXT_${ENTRY_NAME}` constant for easy reuse.
+ * If actors support any specific context entries next to those inherited by the bus action interface,
+ * then this should be described in its README file.
+ *
+ * To avoid entry conflicts, all keys must be properly namespaced using the following convention:
+ *   Each key must be prefixed with the package name followed by a `:`.
+ *   For example, the `rdf-resolve-quad-pattern` bus declares the `sources` entry,
+ *   which should be named as `@comunica/bus-rdf-resolve-quad-pattern:sources`.
+ *
+ * This context can contain any information that might be relevant for certain actors.
+ * For instance, this context can contain a list of datasources over which operators should query.
+ */
+export type ActionContext = Map<string, any>;
+
+export interface IActorInitSparql {
```

Comment:
Since this is a brand new interface, we may consider giving this a more generic name. Perhaps something like `IQueryEngine`?
What do you think?

Reply:
Agreed! And done.